### PR TITLE
fix(sdl): check volume counts to not exceed limits

### PR DESCRIPTION
**Pull request by troian**
Signed-off-by: Artur Troian <troian.ap@gmail.com>

### CODE REVIEW

File: x/deployment/types/v1beta2/resource_list_validation.go [hunk 1]
```diff
@@ -129,6 +129,10 @@ func validateStorage(u types.Volumes) ([]sdk.Int, error) {
 		return nil, errors.Errorf("error: invalid unit storage, cannot be nil")
 	}
 
+	if len(u) > validationConfig.MaxGroupVolumes {
```
Comment (boz) [inline x/deployment/types/v1beta2/resource_list_validation.go:132]: Won't changing this require a chain upgrade?

**Review by boz — CHANGES_REQUESTED** [review]
I think that we should enforce this in the provider, with the max being configurable.

This shouldn't make SDL *parsing* throw an error, tho we can print a warning by validating the manifest or deployment object.

File: x/deployment/types/v1beta2/resource_list_validation.go (hunk 1 above)
Comment (troian) [reply x/deployment/types/v1beta2/resource_list_validation.go:132]: should not as we don't have deployments on both mainnet/testnet with storage.

Comment (boz) [reply x/deployment/types/v1beta2/resource_list_validation.go:132]: but once we do, it will, right?

Comment (troian) [reply x/deployment/types/v1beta2/resource_list_validation.go:132]: correct

Comment (boz) [reply x/deployment/types/v1beta2/resource_list_validation.go:132]: Let's not do that.  I don't want to get into another situation like we were in with the resource limits.

Comment (troian) [reply x/deployment/types/v1beta2/resource_list_validation.go:132]: actually not, it does not change message structure but limits amount of entries

Comment (boz) [reply x/deployment/types/v1beta2/resource_list_validation.go:132]: what if we want to change the limit

Comment (troian) [reply x/deployment/types/v1beta2/resource_list_validation.go:132]: well, yeah, it's gonna be nasty as software upgrade to the network will be required.
i'll make it as limit on the provider side, and print warning as you suggested above

Comment (hydrogen18) [reply x/deployment/types/v1beta2/resource_list_validation.go:132]: Are we just going to put a soft limit in on the provider where it doesn't bid on anything with less than a configured amount of volumes?

Comment (troian) [reply x/deployment/types/v1beta2/resource_list_validation.go:132]: yeah

**Review by boz — APPROVED** [review]
:ship:

File: provider/bidengine/order.go [hunk 2]
```diff
@@ -434,6 +434,12 @@ func (o *order) shouldBid(group *dtypes.Group) (bool, error) {
 		return false, nil
 	}
 
+	for _, resources := range group.GroupSpec.GetResources() {
+		if len(resources.Resources.Storage) > o.cfg.MaxGroupVolumes {
+			o.log.Debug(fmt.Sprintf("unable to fulfill: group volumes count exceeds (%d > %d)", len(resources.Resources.Storage), o.cfg.MaxGroupVolumes))
```
Comment (hydrogen18) [inline provider/bidengine/order.go:439]: Seems like this could just be `Info` log level? It shouldn't happen too often.

**Review by hydrogen18 — APPROVED** [review]

File: provider/bidengine/order.go (hunk 2 above)
Comment (troian) [reply provider/bidengine/order.go:439]: indeed

**Pull request merged**